{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0531c3da",
   "metadata": {},
   "source": [
    "# Attack on scAnnotatR on axilla 10k cells dataset\n",
    "\n",
    "- How to train a scAnnotatR classifier\n",
    "- How to format the classifier to use it with adverSCarial\n",
    "- How to run an IGD4C attack\n",
    "\n",
    "Nguyen, V., Griss, J. scAnnotatR: framework to accurately classify cell types in single-cell RNA-sequencing data. BMC Bioinformatics, 2022. 23(44) https://doi.org/10.1186/s12859-022-04574-5 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1f739233",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: Seurat\n",
      "\n",
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "The legacy packages maptools, rgdal, and rgeos, underpinning the sp package,\n",
      "which was just loaded, will retire in October 2023.\n",
      "Please refer to R-spatial evolution reports for details, especially\n",
      "https://r-spatial.org/r/2023/05/15/evolution4.html.\n",
      "It may be desirable to make the sf package available;\n",
      "package maintainers should consider adding sf to Suggests:.\n",
      "The sp package is now running under evolution status 2\n",
      "     (status 2 uses the sf package in place of rgdal)\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, t\n",
      "\n",
      "\n",
      "Loading required package: SingleCellExperiment\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SeuratObject’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:sp’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘SummarizedExperiment’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Seurat’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SeuratObject’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "Warning message:\n",
      "“replacing previous import ‘ape::where’ by ‘dplyr::where’ when loading ‘scAnnotatR’”\n"
     ]
    }
   ],
   "source": [
    "library(scAnnotatR)\n",
    "library(IRdisplay)\n",
    "library(adverSCarial)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "678bb943",
   "metadata": {},
   "source": [
    "# scAnnotatR classifier training\n",
    "\n",
    "Load previously splitted train/test pbmc3k dataset\n",
    "https://cf.10xgenomics.com/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz\n",
    "\n",
    "annotated from the Seurat tutorial method\n",
    "https://satijalab.org/seurat/articles/pbmc3k_tutorial"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2d9e91a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "c_basen = c(\"hgnc_axilla_10k\", \"hgnc_brain_7k\", \"hgnc_liver_6k\")\n",
    "basen = c_basen[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8ea49383",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "se_train <- readRDS(paste0(\"data//v5/data//sc//\",basen,\"_train.rds\"))\n",
    "se_test <- readRDS(paste0(\"data//v5/data//sc//\",basen,\"_test.rds\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a48c75c0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>malignant cell</li><li>blood vessel endothelial cell</li><li>fibroblast</li><li>macrophage</li><li>blood vessel smooth muscle cell</li><li>T cell</li></ol>\n",
       "\n",
       "<details>\n",
       "\t<summary style=display:list-item;cursor:pointer>\n",
       "\t\t<strong>Levels</strong>:\n",
       "\t</summary>\n",
       "\t<style>\n",
       "\t.list-inline {list-style: none; margin:0; padding: 0}\n",
       "\t.list-inline>li {display: inline-block}\n",
       "\t.list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "\t</style>\n",
       "\t<ol class=list-inline><li>'malignant cell'</li><li>'blood vessel endothelial cell'</li><li>'fibroblast'</li><li>'blood vessel smooth muscle cell'</li><li>'macrophage'</li><li>'T cell'</li></ol>\n",
       "</details>"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item malignant cell\n",
       "\\item blood vessel endothelial cell\n",
       "\\item fibroblast\n",
       "\\item macrophage\n",
       "\\item blood vessel smooth muscle cell\n",
       "\\item T cell\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\emph{Levels}: \\begin{enumerate*}\n",
       "\\item 'malignant cell'\n",
       "\\item 'blood vessel endothelial cell'\n",
       "\\item 'fibroblast'\n",
       "\\item 'blood vessel smooth muscle cell'\n",
       "\\item 'macrophage'\n",
       "\\item 'T cell'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. malignant cell\n",
       "2. blood vessel endothelial cell\n",
       "3. fibroblast\n",
       "4. macrophage\n",
       "5. blood vessel smooth muscle cell\n",
       "6. T cell\n",
       "\n",
       "\n",
       "\n",
       "**Levels**: 1. 'malignant cell'\n",
       "2. 'blood vessel endothelial cell'\n",
       "3. 'fibroblast'\n",
       "4. 'blood vessel smooth muscle cell'\n",
       "5. 'macrophage'\n",
       "6. 'T cell'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] malignant cell                  blood vessel endothelial cell  \n",
       "[3] fibroblast                      macrophage                     \n",
       "[5] blood vessel smooth muscle cell T cell                         \n",
       "6 Levels: malignant cell < blood vessel endothelial cell < ... < T cell"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(se_train@meta.data[['chr_seurat_cluster']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38c3e5a4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b9def834",
   "metadata": {},
   "outputs": [],
   "source": [
    "Idents(se_train) <- \"chr_seurat_cluster\"\n",
    "Idents(se_test) <- \"chr_seurat_cluster\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8fdd272a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 40</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>orig.ident</th><th scope=col>nCount_RNA</th><th scope=col>nFeature_RNA</th><th scope=col>replicate</th><th scope=col>condition</th><th scope=col>labels_unif</th><th scope=col>labels_cl_unif</th><th scope=col>labels_cl_unif2_broad</th><th scope=col>compartments</th><th scope=col>cnv_pass_mal</th><th scope=col>⋯</th><th scope=col>cell_type</th><th scope=col>assay</th><th scope=col>disease</th><th scope=col>organism</th><th scope=col>sex</th><th scope=col>tissue</th><th scope=col>self_reported_ethnicity</th><th scope=col>development_stage</th><th scope=col>observation_joinid</th><th scope=col>chr_seurat_cluster</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>⋯</th><th scope=col>&lt;ord&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;ord&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>HTAPP-878-SMP-7149-TST-channel1_CGTTCTGCAGTCGCTG-1</th><td>HTAPP-878-SMP-7149-TST-channel1</td><td>19735</td><td>6278</td><td>1</td><td>TST</td><td>Epithelial</td><td>Epithelial</td><td>Epithelial</td><td>Malignant</td><td>True</td><td>⋯</td><td>malignant cell</td><td>10x 3' v3</td><td>breast cancer</td><td>Homo sapiens</td><td>female</td><td>axilla</td><td>European</td><td>37-year-old human stage</td><td>KYeaGobvIQ</td><td>malignant cell</td></tr>\n",
       "\t<tr><th scope=row>HTAPP-878-SMP-7149-TST-channel1_GCACGGTTCGATACTG-1</th><td>HTAPP-878-SMP-7149-TST-channel1</td><td>19848</td><td>7140</td><td>1</td><td>TST</td><td>Epithelial</td><td>Epithelial</td><td>Epithelial</td><td>Malignant</td><td>True</td><td>⋯</td><td>malignant cell</td><td>10x 3' v3</td><td>breast cancer</td><td>Homo sapiens</td><td>female</td><td>axilla</td><td>European</td><td>37-year-old human stage</td><td>@dk&lt;quR-I{</td><td>malignant cell</td></tr>\n",
       "\t<tr><th scope=row>HTAPP-878-SMP-7149-TST-channel1_TGTCCACAGTCATAGA-1</th><td>HTAPP-878-SMP-7149-TST-channel1</td><td>19288</td><td>5699</td><td>1</td><td>TST</td><td>Epithelial</td><td>Epithelial</td><td>Epithelial</td><td>Malignant</td><td>True</td><td>⋯</td><td>malignant cell</td><td>10x 3' v3</td><td>breast cancer</td><td>Homo sapiens</td><td>female</td><td>axilla</td><td>European</td><td>37-year-old human stage</td><td>g%ROJ`3nnu</td><td>malignant cell</td></tr>\n",
       "\t<tr><th scope=row>HTAPP-878-SMP-7149-TST-channel1_TTTAGTCTCCACGTCT-1</th><td>HTAPP-878-SMP-7149-TST-channel1</td><td>19569</td><td>6949</td><td>1</td><td>TST</td><td>Epithelial</td><td>Epithelial</td><td>Epithelial</td><td>Malignant</td><td>True</td><td>⋯</td><td>malignant cell</td><td>10x 3' v3</td><td>breast cancer</td><td>Homo sapiens</td><td>female</td><td>axilla</td><td>European</td><td>37-year-old human stage</td><td>(HRUf-FqnM</td><td>malignant cell</td></tr>\n",
       "\t<tr><th scope=row>HTAPP-878-SMP-7149-TST-channel1_CCCGGAACAAGTGGAC-1</th><td>HTAPP-878-SMP-7149-TST-channel1</td><td>19440</td><td>6553</td><td>1</td><td>TST</td><td>Epithelial</td><td>Epithelial</td><td>Epithelial</td><td>Malignant</td><td>True</td><td>⋯</td><td>malignant cell</td><td>10x 3' v3</td><td>breast cancer</td><td>Homo sapiens</td><td>female</td><td>axilla</td><td>European</td><td>37-year-old human stage</td><td>X_LhOX=$9;</td><td>malignant cell</td></tr>\n",
       "\t<tr><th scope=row>HTAPP-878-SMP-7149-TST-channel1_GAGACTTTCATTTCCA-1</th><td>HTAPP-878-SMP-7149-TST-channel1</td><td>19421</td><td>7104</td><td>1</td><td>TST</td><td>Epithelial</td><td>Epithelial</td><td>Epithelial</td><td>Malignant</td><td>True</td><td>⋯</td><td>malignant cell</td><td>10x 3' v3</td><td>breast cancer</td><td>Homo sapiens</td><td>female</td><td>axilla</td><td>European</td><td>37-year-old human stage</td><td>r0fjWR$I_}</td><td>malignant cell</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 40\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & orig.ident & nCount\\_RNA & nFeature\\_RNA & replicate & condition & labels\\_unif & labels\\_cl\\_unif & labels\\_cl\\_unif2\\_broad & compartments & cnv\\_pass\\_mal & ⋯ & cell\\_type & assay & disease & organism & sex & tissue & self\\_reported\\_ethnicity & development\\_stage & observation\\_joinid & chr\\_seurat\\_cluster\\\\\n",
       "  & <fct> & <dbl> & <int> & <fct> & <fct> & <fct> & <fct> & <fct> & <fct> & <fct> & ⋯ & <ord> & <fct> & <fct> & <fct> & <fct> & <fct> & <fct> & <fct> & <chr> & <ord>\\\\\n",
       "\\hline\n",
       "\tHTAPP-878-SMP-7149-TST-channel1\\_CGTTCTGCAGTCGCTG-1 & HTAPP-878-SMP-7149-TST-channel1 & 19735 & 6278 & 1 & TST & Epithelial & Epithelial & Epithelial & Malignant & True & ⋯ & malignant cell & 10x 3' v3 & breast cancer & Homo sapiens & female & axilla & European & 37-year-old human stage & KYeaGobvIQ & malignant cell\\\\\n",
       "\tHTAPP-878-SMP-7149-TST-channel1\\_GCACGGTTCGATACTG-1 & HTAPP-878-SMP-7149-TST-channel1 & 19848 & 7140 & 1 & TST & Epithelial & Epithelial & Epithelial & Malignant & True & ⋯ & malignant cell & 10x 3' v3 & breast cancer & Homo sapiens & female & axilla & European & 37-year-old human stage & @dk<quR-I\\{ & malignant cell\\\\\n",
       "\tHTAPP-878-SMP-7149-TST-channel1\\_TGTCCACAGTCATAGA-1 & HTAPP-878-SMP-7149-TST-channel1 & 19288 & 5699 & 1 & TST & Epithelial & Epithelial & Epithelial & Malignant & True & ⋯ & malignant cell & 10x 3' v3 & breast cancer & Homo sapiens & female & axilla & European & 37-year-old human stage & g\\%ROJ`3nnu & malignant cell\\\\\n",
       "\tHTAPP-878-SMP-7149-TST-channel1\\_TTTAGTCTCCACGTCT-1 & HTAPP-878-SMP-7149-TST-channel1 & 19569 & 6949 & 1 & TST & Epithelial & Epithelial & Epithelial & Malignant & True & ⋯ & malignant cell & 10x 3' v3 & breast cancer & Homo sapiens & female & axilla & European & 37-year-old human stage & (HRUf-FqnM & malignant cell\\\\\n",
       "\tHTAPP-878-SMP-7149-TST-channel1\\_CCCGGAACAAGTGGAC-1 & HTAPP-878-SMP-7149-TST-channel1 & 19440 & 6553 & 1 & TST & Epithelial & Epithelial & Epithelial & Malignant & True & ⋯ & malignant cell & 10x 3' v3 & breast cancer & Homo sapiens & female & axilla & European & 37-year-old human stage & X\\_LhOX=\\$9; & malignant cell\\\\\n",
       "\tHTAPP-878-SMP-7149-TST-channel1\\_GAGACTTTCATTTCCA-1 & HTAPP-878-SMP-7149-TST-channel1 & 19421 & 7104 & 1 & TST & Epithelial & Epithelial & Epithelial & Malignant & True & ⋯ & malignant cell & 10x 3' v3 & breast cancer & Homo sapiens & female & axilla & European & 37-year-old human stage & r0fjWR\\$I\\_\\} & malignant cell\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 40\n",
       "\n",
       "| <!--/--> | orig.ident &lt;fct&gt; | nCount_RNA &lt;dbl&gt; | nFeature_RNA &lt;int&gt; | replicate &lt;fct&gt; | condition &lt;fct&gt; | labels_unif &lt;fct&gt; | labels_cl_unif &lt;fct&gt; | labels_cl_unif2_broad &lt;fct&gt; | compartments &lt;fct&gt; | cnv_pass_mal &lt;fct&gt; | ⋯ ⋯ | cell_type &lt;ord&gt; | assay &lt;fct&gt; | disease &lt;fct&gt; | organism &lt;fct&gt; | sex &lt;fct&gt; | tissue &lt;fct&gt; | self_reported_ethnicity &lt;fct&gt; | development_stage &lt;fct&gt; | observation_joinid &lt;chr&gt; | chr_seurat_cluster &lt;ord&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| HTAPP-878-SMP-7149-TST-channel1_CGTTCTGCAGTCGCTG-1 | HTAPP-878-SMP-7149-TST-channel1 | 19735 | 6278 | 1 | TST | Epithelial | Epithelial | Epithelial | Malignant | True | ⋯ | malignant cell | 10x 3' v3 | breast cancer | Homo sapiens | female | axilla | European | 37-year-old human stage | KYeaGobvIQ | malignant cell |\n",
       "| HTAPP-878-SMP-7149-TST-channel1_GCACGGTTCGATACTG-1 | HTAPP-878-SMP-7149-TST-channel1 | 19848 | 7140 | 1 | TST | Epithelial | Epithelial | Epithelial | Malignant | True | ⋯ | malignant cell | 10x 3' v3 | breast cancer | Homo sapiens | female | axilla | European | 37-year-old human stage | @dk&lt;quR-I{ | malignant cell |\n",
       "| HTAPP-878-SMP-7149-TST-channel1_TGTCCACAGTCATAGA-1 | HTAPP-878-SMP-7149-TST-channel1 | 19288 | 5699 | 1 | TST | Epithelial | Epithelial | Epithelial | Malignant | True | ⋯ | malignant cell | 10x 3' v3 | breast cancer | Homo sapiens | female | axilla | European | 37-year-old human stage | g%ROJ`3nnu | malignant cell |\n",
       "| HTAPP-878-SMP-7149-TST-channel1_TTTAGTCTCCACGTCT-1 | HTAPP-878-SMP-7149-TST-channel1 | 19569 | 6949 | 1 | TST | Epithelial | Epithelial | Epithelial | Malignant | True | ⋯ | malignant cell | 10x 3' v3 | breast cancer | Homo sapiens | female | axilla | European | 37-year-old human stage | (HRUf-FqnM | malignant cell |\n",
       "| HTAPP-878-SMP-7149-TST-channel1_CCCGGAACAAGTGGAC-1 | HTAPP-878-SMP-7149-TST-channel1 | 19440 | 6553 | 1 | TST | Epithelial | Epithelial | Epithelial | Malignant | True | ⋯ | malignant cell | 10x 3' v3 | breast cancer | Homo sapiens | female | axilla | European | 37-year-old human stage | X_LhOX=$9; | malignant cell |\n",
       "| HTAPP-878-SMP-7149-TST-channel1_GAGACTTTCATTTCCA-1 | HTAPP-878-SMP-7149-TST-channel1 | 19421 | 7104 | 1 | TST | Epithelial | Epithelial | Epithelial | Malignant | True | ⋯ | malignant cell | 10x 3' v3 | breast cancer | Homo sapiens | female | axilla | European | 37-year-old human stage | r0fjWR$I_} | malignant cell |\n",
       "\n"
      ],
      "text/plain": [
       "                                                   orig.ident                     \n",
       "HTAPP-878-SMP-7149-TST-channel1_CGTTCTGCAGTCGCTG-1 HTAPP-878-SMP-7149-TST-channel1\n",
       "HTAPP-878-SMP-7149-TST-channel1_GCACGGTTCGATACTG-1 HTAPP-878-SMP-7149-TST-channel1\n",
       "HTAPP-878-SMP-7149-TST-channel1_TGTCCACAGTCATAGA-1 HTAPP-878-SMP-7149-TST-channel1\n",
       "HTAPP-878-SMP-7149-TST-channel1_TTTAGTCTCCACGTCT-1 HTAPP-878-SMP-7149-TST-channel1\n",
       "HTAPP-878-SMP-7149-TST-channel1_CCCGGAACAAGTGGAC-1 HTAPP-878-SMP-7149-TST-channel1\n",
       "HTAPP-878-SMP-7149-TST-channel1_GAGACTTTCATTTCCA-1 HTAPP-878-SMP-7149-TST-channel1\n",
       "                                                   nCount_RNA nFeature_RNA\n",
       "HTAPP-878-SMP-7149-TST-channel1_CGTTCTGCAGTCGCTG-1 19735      6278        \n",
       "HTAPP-878-SMP-7149-TST-channel1_GCACGGTTCGATACTG-1 19848      7140        \n",
       "HTAPP-878-SMP-7149-TST-channel1_TGTCCACAGTCATAGA-1 19288      5699        \n",
       "HTAPP-878-SMP-7149-TST-channel1_TTTAGTCTCCACGTCT-1 19569      6949        \n",
       "HTAPP-878-SMP-7149-TST-channel1_CCCGGAACAAGTGGAC-1 19440      6553        \n",
       "HTAPP-878-SMP-7149-TST-channel1_GAGACTTTCATTTCCA-1 19421      7104        \n",
       "                                                   replicate condition\n",
       "HTAPP-878-SMP-7149-TST-channel1_CGTTCTGCAGTCGCTG-1 1         TST      \n",
       "HTAPP-878-SMP-7149-TST-channel1_GCACGGTTCGATACTG-1 1         TST      \n",
       "HTAPP-878-SMP-7149-TST-channel1_TGTCCACAGTCATAGA-1 1         TST      \n",
       "HTAPP-878-SMP-7149-TST-channel1_TTTAGTCTCCACGTCT-1 1         TST      \n",
       "HTAPP-878-SMP-7149-TST-channel1_CCCGGAACAAGTGGAC-1 1         TST      \n",
       "HTAPP-878-SMP-7149-TST-channel1_GAGACTTTCATTTCCA-1 1         TST      \n",
       "                                                   labels_unif labels_cl_unif\n",
       "HTAPP-878-SMP-7149-TST-channel1_CGTTCTGCAGTCGCTG-1 Epithelial  Epithelial    \n",
       "HTAPP-878-SMP-7149-TST-channel1_GCACGGTTCGATACTG-1 Epithelial  Epithelial    \n",
       "HTAPP-878-SMP-7149-TST-channel1_TGTCCACAGTCATAGA-1 Epithelial  Epithelial    \n",
       "HTAPP-878-SMP-7149-TST-channel1_TTTAGTCTCCACGTCT-1 Epithelial  Epithelial    \n",
       "HTAPP-878-SMP-7149-TST-channel1_CCCGGAACAAGTGGAC-1 Epithelial  Epithelial    \n",
       "HTAPP-878-SMP-7149-TST-channel1_GAGACTTTCATTTCCA-1 Epithelial  Epithelial    \n",
       "                                                   labels_cl_unif2_broad\n",
       "HTAPP-878-SMP-7149-TST-channel1_CGTTCTGCAGTCGCTG-1 Epithelial           \n",
       "HTAPP-878-SMP-7149-TST-channel1_GCACGGTTCGATACTG-1 Epithelial           \n",
       "HTAPP-878-SMP-7149-TST-channel1_TGTCCACAGTCATAGA-1 Epithelial           \n",
       "HTAPP-878-SMP-7149-TST-channel1_TTTAGTCTCCACGTCT-1 Epithelial           \n",
       "HTAPP-878-SMP-7149-TST-channel1_CCCGGAACAAGTGGAC-1 Epithelial           \n",
       "HTAPP-878-SMP-7149-TST-channel1_GAGACTTTCATTTCCA-1 Epithelial           \n",
       "                                                   compartments cnv_pass_mal ⋯\n",
       "HTAPP-878-SMP-7149-TST-channel1_CGTTCTGCAGTCGCTG-1 Malignant    True         ⋯\n",
       "HTAPP-878-SMP-7149-TST-channel1_GCACGGTTCGATACTG-1 Malignant    True         ⋯\n",
       "HTAPP-878-SMP-7149-TST-channel1_TGTCCACAGTCATAGA-1 Malignant    True         ⋯\n",
       "HTAPP-878-SMP-7149-TST-channel1_TTTAGTCTCCACGTCT-1 Malignant    True         ⋯\n",
       "HTAPP-878-SMP-7149-TST-channel1_CCCGGAACAAGTGGAC-1 Malignant    True         ⋯\n",
       "HTAPP-878-SMP-7149-TST-channel1_GAGACTTTCATTTCCA-1 Malignant    True         ⋯\n",
       "                                                   cell_type      assay    \n",
       "HTAPP-878-SMP-7149-TST-channel1_CGTTCTGCAGTCGCTG-1 malignant cell 10x 3' v3\n",
       "HTAPP-878-SMP-7149-TST-channel1_GCACGGTTCGATACTG-1 malignant cell 10x 3' v3\n",
       "HTAPP-878-SMP-7149-TST-channel1_TGTCCACAGTCATAGA-1 malignant cell 10x 3' v3\n",
       "HTAPP-878-SMP-7149-TST-channel1_TTTAGTCTCCACGTCT-1 malignant cell 10x 3' v3\n",
       "HTAPP-878-SMP-7149-TST-channel1_CCCGGAACAAGTGGAC-1 malignant cell 10x 3' v3\n",
       "HTAPP-878-SMP-7149-TST-channel1_GAGACTTTCATTTCCA-1 malignant cell 10x 3' v3\n",
       "                                                   disease       organism    \n",
       "HTAPP-878-SMP-7149-TST-channel1_CGTTCTGCAGTCGCTG-1 breast cancer Homo sapiens\n",
       "HTAPP-878-SMP-7149-TST-channel1_GCACGGTTCGATACTG-1 breast cancer Homo sapiens\n",
       "HTAPP-878-SMP-7149-TST-channel1_TGTCCACAGTCATAGA-1 breast cancer Homo sapiens\n",
       "HTAPP-878-SMP-7149-TST-channel1_TTTAGTCTCCACGTCT-1 breast cancer Homo sapiens\n",
       "HTAPP-878-SMP-7149-TST-channel1_CCCGGAACAAGTGGAC-1 breast cancer Homo sapiens\n",
       "HTAPP-878-SMP-7149-TST-channel1_GAGACTTTCATTTCCA-1 breast cancer Homo sapiens\n",
       "                                                   sex    tissue\n",
       "HTAPP-878-SMP-7149-TST-channel1_CGTTCTGCAGTCGCTG-1 female axilla\n",
       "HTAPP-878-SMP-7149-TST-channel1_GCACGGTTCGATACTG-1 female axilla\n",
       "HTAPP-878-SMP-7149-TST-channel1_TGTCCACAGTCATAGA-1 female axilla\n",
       "HTAPP-878-SMP-7149-TST-channel1_TTTAGTCTCCACGTCT-1 female axilla\n",
       "HTAPP-878-SMP-7149-TST-channel1_CCCGGAACAAGTGGAC-1 female axilla\n",
       "HTAPP-878-SMP-7149-TST-channel1_GAGACTTTCATTTCCA-1 female axilla\n",
       "                                                   self_reported_ethnicity\n",
       "HTAPP-878-SMP-7149-TST-channel1_CGTTCTGCAGTCGCTG-1 European               \n",
       "HTAPP-878-SMP-7149-TST-channel1_GCACGGTTCGATACTG-1 European               \n",
       "HTAPP-878-SMP-7149-TST-channel1_TGTCCACAGTCATAGA-1 European               \n",
       "HTAPP-878-SMP-7149-TST-channel1_TTTAGTCTCCACGTCT-1 European               \n",
       "HTAPP-878-SMP-7149-TST-channel1_CCCGGAACAAGTGGAC-1 European               \n",
       "HTAPP-878-SMP-7149-TST-channel1_GAGACTTTCATTTCCA-1 European               \n",
       "                                                   development_stage      \n",
       "HTAPP-878-SMP-7149-TST-channel1_CGTTCTGCAGTCGCTG-1 37-year-old human stage\n",
       "HTAPP-878-SMP-7149-TST-channel1_GCACGGTTCGATACTG-1 37-year-old human stage\n",
       "HTAPP-878-SMP-7149-TST-channel1_TGTCCACAGTCATAGA-1 37-year-old human stage\n",
       "HTAPP-878-SMP-7149-TST-channel1_TTTAGTCTCCACGTCT-1 37-year-old human stage\n",
       "HTAPP-878-SMP-7149-TST-channel1_CCCGGAACAAGTGGAC-1 37-year-old human stage\n",
       "HTAPP-878-SMP-7149-TST-channel1_GAGACTTTCATTTCCA-1 37-year-old human stage\n",
       "                                                   observation_joinid\n",
       "HTAPP-878-SMP-7149-TST-channel1_CGTTCTGCAGTCGCTG-1 KYeaGobvIQ        \n",
       "HTAPP-878-SMP-7149-TST-channel1_GCACGGTTCGATACTG-1 @dk<quR-I{        \n",
       "HTAPP-878-SMP-7149-TST-channel1_TGTCCACAGTCATAGA-1 g%ROJ`3nnu        \n",
       "HTAPP-878-SMP-7149-TST-channel1_TTTAGTCTCCACGTCT-1 (HRUf-FqnM        \n",
       "HTAPP-878-SMP-7149-TST-channel1_CCCGGAACAAGTGGAC-1 X_LhOX=$9;        \n",
       "HTAPP-878-SMP-7149-TST-channel1_GAGACTTTCATTTCCA-1 r0fjWR$I_}        \n",
       "                                                   chr_seurat_cluster\n",
       "HTAPP-878-SMP-7149-TST-channel1_CGTTCTGCAGTCGCTG-1 malignant cell    \n",
       "HTAPP-878-SMP-7149-TST-channel1_GCACGGTTCGATACTG-1 malignant cell    \n",
       "HTAPP-878-SMP-7149-TST-channel1_TGTCCACAGTCATAGA-1 malignant cell    \n",
       "HTAPP-878-SMP-7149-TST-channel1_TTTAGTCTCCACGTCT-1 malignant cell    \n",
       "HTAPP-878-SMP-7149-TST-channel1_CCCGGAACAAGTGGAC-1 malignant cell    \n",
       "HTAPP-878-SMP-7149-TST-channel1_GAGACTTTCATTTCCA-1 malignant cell    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(se_train@meta.data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b8f3d1f",
   "metadata": {},
   "source": [
    "## Function to get the most significant genes between a cluster and the other cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "46db0599",
   "metadata": {},
   "outputs": [],
   "source": [
    "getSignGenesNot <- function(expr, clusters, target, method=\"wilcox\", adjMethod=\"BH\", verbose=FALSE){\n",
    "        if (verbose) {message(\"Cluster \",target,\" vs all the other clusters\")}\n",
    "        pvals <- apply(t(expr), 1, function(x){\n",
    "            c1 <- x[clusters == target]\n",
    "            c1 <- c1[!is.na(c1)]\n",
    "            c2 <- x[clusters != target]\n",
    "            c2 <- c2[!is.na(c2)]\n",
    "            if ( length(c1) == 0 || length(c2) == 0){\n",
    "                return(1)\n",
    "            }\n",
    "            if (length(unique(c1))==1){\n",
    "                c1[1] = c1[1] + 0.00001\n",
    "            }\n",
    "            if (length(unique(c2))==1){\n",
    "                c2[1] = c2[1] + 0.00001\n",
    "            }\n",
    "            if (method==\"wilcox\"){\n",
    "                return(wilcox.test(c1, c2)$p.value)\n",
    "            }\n",
    "            if (method==\"ttest\"){\n",
    "                return(t.test(c1, c2)$p.value)\n",
    "            }\n",
    "        })\n",
    "        means <- apply(t(expr), 1, function(x){\n",
    "                c1 <- x[clusters == target]\n",
    "                c1 <- c1[!is.na(c1)]\n",
    "                c2 <- x[clusters != target]\n",
    "                c2 <- c2[!is.na(c2)]\n",
    "                if (length(c1) == 0){\n",
    "                    c1 = c(0)\n",
    "                }\n",
    "                if (length(c2) == 0){\n",
    "                    c2 = c(0)\n",
    "                }\n",
    "                return(mean(c1)-mean(c2))\n",
    "        })  \n",
    "        dfPvals <- data.frame(gene=colnames(expr), pval=unname(pvals), mean=means)\n",
    "        rownames(dfPvals) <- dfPvals$gene\n",
    "\n",
    "        for (clustInt in setdiff(unique(clusters), target)){\n",
    "            if(verbose){message(\"Cluster \", target, \" vs cluster \", clustInt)}\n",
    "            newPvals <- pvals <- apply(t(expr), 1, function(x){\n",
    "                c1 <- x[clusters == target]\n",
    "                c1 <- c1[!is.na(c1)]\n",
    "                c2 <- x[clusters == clustInt]\n",
    "                c2 <- c2[!is.na(c2)]\n",
    "                if ( length(c1) == 0 || length(c2) == 0){\n",
    "                    return(1)\n",
    "                }\n",
    "                if (length(unique(c1))==1){\n",
    "                    c1[1] = c1[1] + 0.001\n",
    "                }\n",
    "                if (length(unique(c2))==1){\n",
    "                    c2[1] = c2[1] + 0.001\n",
    "                }\n",
    "                if (method==\"wilcox\"){\n",
    "                    return(wilcox.test(c1, c2)$p.value)\n",
    "                }\n",
    "                if (method==\"ttest\"){\n",
    "                    return(t.test(c1, c2)$p.value)\n",
    "                }    \n",
    "            })\n",
    "            newMeans <- pvals <- apply(t(expr), 1, function(x){\n",
    "                c1 <- x[clusters == target]\n",
    "                c1 <- c1[!is.na(c1)]\n",
    "                c2 <- x[clusters == clustInt]\n",
    "                c2 <- c2[!is.na(c2)]\n",
    "                if (length(c1) == 0){\n",
    "                    c1 = c(0)\n",
    "                }\n",
    "                if (length(c2) == 0){\n",
    "                    c2 = c(0)\n",
    "                }\n",
    "                return(mean(c1)-mean(c2))\n",
    "            })  \n",
    "            if(verbose){message(sum(unname(newPvals) < dfPvals$pval),\" pvalues replaced by lower values\")}\n",
    "            dfPvals[unname(newPvals) < dfPvals$pval, \"pval\"] <- newPvals[unname(newPvals) < dfPvals$pval]\n",
    "            dfPvals[unname(newPvals) < dfPvals$pval, \"mean\"] <- newMeans[unname(newPvals) < dfPvals$pval]\n",
    "    }\n",
    "    dfPvals$adjPval <- p.adjust(dfPvals$pval, method=adjMethod)\n",
    "    dfPvals <- dfPvals[order(dfPvals$pval),]\n",
    "    return(dfPvals)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "199e5725",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "12ce1a1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "dfScaled <- as.data.frame(se_train@assays$RNA@layers$scale.data)\n",
    "colnames(dfScaled) <- rownames(se_train@assays$RNA@cells)\n",
    "rownames(dfScaled) <- rownames(se_train@assays$RNA@features)\n",
    "dfScaled <- as.data.frame(t(dfScaled))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c63941ae",
   "metadata": {},
   "source": [
    "The scAnnotatR will build one classifier to predict each cell type, so we one-hot encode the cell type by adding a meta.data column for each cell type with binary information: cell type or \"unknown\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "74311bb0",
   "metadata": {},
   "outputs": [],
   "source": [
    "for (cellType in unique(se_train@meta.data$chr_seurat_cluster)){\n",
    "    se_train@meta.data[[cellType]] <- unlist(lapply(se_train@meta.data$chr_seurat_cluster, function(x){\n",
    "        if (x == cellType){\n",
    "            return(cellType)\n",
    "        } else {\n",
    "            return(\"unknown\")\n",
    "        }\n",
    "    }))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1945da64",
   "metadata": {},
   "source": [
    "## Train and export one classifier for each cell type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7bfd0f08",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'malignant cell'"
      ],
      "text/latex": [
       "'malignant cell'"
      ],
      "text/markdown": [
       "'malignant cell'"
      ],
      "text/plain": [
       "[1] \"malignant cell\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "listClassifiers <- list()\n",
    "for (cellType in unique(se_train@meta.data$chr_seurat_cluster)){\n",
    "    display(cellType)\n",
    "    sg <- getSignGenesNot(dfScaled,\n",
    "                   se_train@meta.data$chr_seurat_cluster,\n",
    "                   cellType)\n",
    "    # Selection of the 20 most significants with a minimum mean difference of 0.5 for the SVM\n",
    "    selectedMarkers <- rownames(sg[abs(sg$mean)>0.5,])[1:20]\n",
    "    classifier <- train_classifier(train_obj = se_train, cell_type = cellType, \n",
    "                                 marker_genes = selectedMarkers,\n",
    "                                 assay = 'RNA', tag_slot = cellType)\n",
    "    listClassifiers[[cellType]] <- classifier\n",
    "    \n",
    "    save_new_model(new_model = classifier,\n",
    "                   path_to_models = paste0(\"repr_data/classifiers/scAnnotatR/trainedClass_\", basen),\n",
    "               include.default = FALSE) \n",
    "}\n",
    "listClassifiers[[cellType]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1522de4",
   "metadata": {},
   "source": [
    "# Format the Classifier\n",
    "To work with adverSCarial the classifier needs to be formated in a certain way."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "75460d09",
   "metadata": {},
   "outputs": [],
   "source": [
    "scAnnotatR_classifier = function(expr, clusters, target){\n",
    "    if (!\"scAnnotatR\" %in% loadedNamespaces()){\n",
    "        library(scAnnotatR)\n",
    "    }\n",
    "    if ( !exists(\"sca_default_models\")){\n",
    "        sca_default_models <<- load_models(paste0(\"repr_data/classifiers/scAnnotatR/trainedClass_\", basen))\n",
    "        sca_cell_types <<- names(sca_default_models)\n",
    "    }\n",
    "    \n",
    "    seurat.obj <- classify_cells(classify_obj = expr, \n",
    "                             assay = 'RNA', slot = 'scale.data',\n",
    "                             cell_types = sca_cell_types, \n",
    "                             path_to_models = paste0(\"repr_data/classifiers/scAnnotatR/trainedClass_\", basen))\n",
    "    typePredictions <- seurat.obj@meta.data[, stringr::str_replace_all( paste0(sca_cell_types, \"_p\"), \" \", \"_\")]\n",
    "    \n",
    "    pred_table <- table(seurat.obj@meta.data[clusters == target, \"most_probable_cell_type\"])\n",
    "    cluster_pred <- names(pred_table[order(pred_table, decreasing = TRUE)])[1]\n",
    "    odd <- unname(pred_table[cluster_pred]/sum(pred_table))\n",
    "    colnames(typePredictions) <- unlist(lapply(colnames(typePredictions), function(x){\n",
    "        unlist(strsplit(x,\"_p$\"))[1]\n",
    "    }))\n",
    "    typePredictions <- as.data.frame(t(typePredictions))\n",
    "    \n",
    "    result <- list(\n",
    "        # Cell type prediction for the cluster\n",
    "        prediction=stringr::str_replace_all(cluster_pred,\" \",\"_\"),\n",
    "        # Score of the predicted cell type\n",
    "        odd=odd, \n",
    "        # Score for each cell type for each cell\n",
    "        typePredictions=typePredictions,\n",
    "        # Cell type for each cell\n",
    "        cellTypes=seurat.obj@meta.data$most_probable_cell_type)\n",
    "    return(result)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41329ee4",
   "metadata": {},
   "source": [
    "# Adversarial attack of the malignant cell cells cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5fe01eb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_pbmc_test <- se_test\n",
    "clusters_so = so_pbmc_test@meta.data$chr_seurat_cluster\n",
    "names(clusters_so) <- rownames(so_pbmc_test@meta.data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b505ee93",
   "metadata": {},
   "source": [
    "cell types clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c6210693",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>malignant cell</li><li>blood vessel endothelial cell</li><li>fibroblast</li><li>macrophage</li><li>blood vessel smooth muscle cell</li><li>T cell</li></ol>\n",
       "\n",
       "<details>\n",
       "\t<summary style=display:list-item;cursor:pointer>\n",
       "\t\t<strong>Levels</strong>:\n",
       "\t</summary>\n",
       "\t<style>\n",
       "\t.list-inline {list-style: none; margin:0; padding: 0}\n",
       "\t.list-inline>li {display: inline-block}\n",
       "\t.list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "\t</style>\n",
       "\t<ol class=list-inline><li>'malignant cell'</li><li>'blood vessel endothelial cell'</li><li>'fibroblast'</li><li>'blood vessel smooth muscle cell'</li><li>'macrophage'</li><li>'T cell'</li></ol>\n",
       "</details>"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item malignant cell\n",
       "\\item blood vessel endothelial cell\n",
       "\\item fibroblast\n",
       "\\item macrophage\n",
       "\\item blood vessel smooth muscle cell\n",
       "\\item T cell\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\emph{Levels}: \\begin{enumerate*}\n",
       "\\item 'malignant cell'\n",
       "\\item 'blood vessel endothelial cell'\n",
       "\\item 'fibroblast'\n",
       "\\item 'blood vessel smooth muscle cell'\n",
       "\\item 'macrophage'\n",
       "\\item 'T cell'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. malignant cell\n",
       "2. blood vessel endothelial cell\n",
       "3. fibroblast\n",
       "4. macrophage\n",
       "5. blood vessel smooth muscle cell\n",
       "6. T cell\n",
       "\n",
       "\n",
       "\n",
       "**Levels**: 1. 'malignant cell'\n",
       "2. 'blood vessel endothelial cell'\n",
       "3. 'fibroblast'\n",
       "4. 'blood vessel smooth muscle cell'\n",
       "5. 'macrophage'\n",
       "6. 'T cell'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] malignant cell                  blood vessel endothelial cell  \n",
       "[3] fibroblast                      macrophage                     \n",
       "[5] blood vessel smooth muscle cell T cell                         \n",
       "6 Levels: malignant cell < blood vessel endothelial cell < ... < T cell"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(clusters_so)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d92d4f6",
   "metadata": {},
   "source": [
    "We check if the classifier is working properly, it predicts the NK cells cluster as NK cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c0243451",
   "metadata": {},
   "outputs": [],
   "source": [
    "class_results <- scAnnotatR_classifier(so_pbmc_test, clusters_so, 'macrophage')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "bc6e9e4f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'macrophage'"
      ],
      "text/latex": [
       "'macrophage'"
      ],
      "text/markdown": [
       "'macrophage'"
      ],
      "text/plain": [
       "[1] \"macrophage\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "class_results$prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e88fd8c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "dfScaled <- as.data.frame(so_pbmc_test@assays$RNA@layers$scale.data)\n",
    "colnames(dfScaled) <- rownames(so_pbmc_test@assays$RNA@cells)\n",
    "rownames(dfScaled) <- rownames(so_pbmc_test@assays$RNA@features)\n",
    "dfScaled <- t(dfScaled)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "211db708",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 5 × 5 of type dbl</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>ENSG00000238009</th><th scope=col>ENSG00000241599</th><th scope=col>ENSG00000235146</th><th scope=col>LINC01409</th><th scope=col>FAM87B</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>HTAPP-878-SMP-7149-TST-channel1_GATGTTGCAAACGTGG-1</th><td>-0.2027578</td><td>-0.01920534</td><td>-0.04245598</td><td>0.7910192</td><td>-0.01895955</td></tr>\n",
       "\t<tr><th scope=row>HTAPP-878-SMP-7149-TST-channel1_CTTTCAAGTAGGTACG-1</th><td>-0.2027578</td><td>-0.01920534</td><td>-0.04245598</td><td>0.2370915</td><td>-0.01895955</td></tr>\n",
       "\t<tr><th scope=row>HTAPP-878-SMP-7149-TST-channel1_CACCGTTGTTCTGACA-1</th><td>-0.2027578</td><td>-0.01920534</td><td>-0.04245598</td><td>0.7840407</td><td>-0.01895955</td></tr>\n",
       "\t<tr><th scope=row>HTAPP-878-SMP-7149-TST-channel1_GAGTTTGCACAACGTT-1</th><td> 1.5589332</td><td>-0.01920534</td><td>-0.04245598</td><td>0.8048589</td><td>-0.01895955</td></tr>\n",
       "\t<tr><th scope=row>HTAPP-878-SMP-7149-TST-channel1_GATGACTTCTTTGCTA-1</th><td>-0.2027578</td><td>-0.01920534</td><td>-0.04245598</td><td>0.2525219</td><td>-0.01895955</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 5 × 5 of type dbl\n",
       "\\begin{tabular}{r|lllll}\n",
       "  & ENSG00000238009 & ENSG00000241599 & ENSG00000235146 & LINC01409 & FAM87B\\\\\n",
       "\\hline\n",
       "\tHTAPP-878-SMP-7149-TST-channel1\\_GATGTTGCAAACGTGG-1 & -0.2027578 & -0.01920534 & -0.04245598 & 0.7910192 & -0.01895955\\\\\n",
       "\tHTAPP-878-SMP-7149-TST-channel1\\_CTTTCAAGTAGGTACG-1 & -0.2027578 & -0.01920534 & -0.04245598 & 0.2370915 & -0.01895955\\\\\n",
       "\tHTAPP-878-SMP-7149-TST-channel1\\_CACCGTTGTTCTGACA-1 & -0.2027578 & -0.01920534 & -0.04245598 & 0.7840407 & -0.01895955\\\\\n",
       "\tHTAPP-878-SMP-7149-TST-channel1\\_GAGTTTGCACAACGTT-1 &  1.5589332 & -0.01920534 & -0.04245598 & 0.8048589 & -0.01895955\\\\\n",
       "\tHTAPP-878-SMP-7149-TST-channel1\\_GATGACTTCTTTGCTA-1 & -0.2027578 & -0.01920534 & -0.04245598 & 0.2525219 & -0.01895955\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 5 × 5 of type dbl\n",
       "\n",
       "| <!--/--> | ENSG00000238009 | ENSG00000241599 | ENSG00000235146 | LINC01409 | FAM87B |\n",
       "|---|---|---|---|---|---|\n",
       "| HTAPP-878-SMP-7149-TST-channel1_GATGTTGCAAACGTGG-1 | -0.2027578 | -0.01920534 | -0.04245598 | 0.7910192 | -0.01895955 |\n",
       "| HTAPP-878-SMP-7149-TST-channel1_CTTTCAAGTAGGTACG-1 | -0.2027578 | -0.01920534 | -0.04245598 | 0.2370915 | -0.01895955 |\n",
       "| HTAPP-878-SMP-7149-TST-channel1_CACCGTTGTTCTGACA-1 | -0.2027578 | -0.01920534 | -0.04245598 | 0.7840407 | -0.01895955 |\n",
       "| HTAPP-878-SMP-7149-TST-channel1_GAGTTTGCACAACGTT-1 |  1.5589332 | -0.01920534 | -0.04245598 | 0.8048589 | -0.01895955 |\n",
       "| HTAPP-878-SMP-7149-TST-channel1_GATGACTTCTTTGCTA-1 | -0.2027578 | -0.01920534 | -0.04245598 | 0.2525219 | -0.01895955 |\n",
       "\n"
      ],
      "text/plain": [
       "                                                   ENSG00000238009\n",
       "HTAPP-878-SMP-7149-TST-channel1_GATGTTGCAAACGTGG-1 -0.2027578     \n",
       "HTAPP-878-SMP-7149-TST-channel1_CTTTCAAGTAGGTACG-1 -0.2027578     \n",
       "HTAPP-878-SMP-7149-TST-channel1_CACCGTTGTTCTGACA-1 -0.2027578     \n",
       "HTAPP-878-SMP-7149-TST-channel1_GAGTTTGCACAACGTT-1  1.5589332     \n",
       "HTAPP-878-SMP-7149-TST-channel1_GATGACTTCTTTGCTA-1 -0.2027578     \n",
       "                                                   ENSG00000241599\n",
       "HTAPP-878-SMP-7149-TST-channel1_GATGTTGCAAACGTGG-1 -0.01920534    \n",
       "HTAPP-878-SMP-7149-TST-channel1_CTTTCAAGTAGGTACG-1 -0.01920534    \n",
       "HTAPP-878-SMP-7149-TST-channel1_CACCGTTGTTCTGACA-1 -0.01920534    \n",
       "HTAPP-878-SMP-7149-TST-channel1_GAGTTTGCACAACGTT-1 -0.01920534    \n",
       "HTAPP-878-SMP-7149-TST-channel1_GATGACTTCTTTGCTA-1 -0.01920534    \n",
       "                                                   ENSG00000235146 LINC01409\n",
       "HTAPP-878-SMP-7149-TST-channel1_GATGTTGCAAACGTGG-1 -0.04245598     0.7910192\n",
       "HTAPP-878-SMP-7149-TST-channel1_CTTTCAAGTAGGTACG-1 -0.04245598     0.2370915\n",
       "HTAPP-878-SMP-7149-TST-channel1_CACCGTTGTTCTGACA-1 -0.04245598     0.7840407\n",
       "HTAPP-878-SMP-7149-TST-channel1_GAGTTTGCACAACGTT-1 -0.04245598     0.8048589\n",
       "HTAPP-878-SMP-7149-TST-channel1_GATGACTTCTTTGCTA-1 -0.04245598     0.2525219\n",
       "                                                   FAM87B     \n",
       "HTAPP-878-SMP-7149-TST-channel1_GATGTTGCAAACGTGG-1 -0.01895955\n",
       "HTAPP-878-SMP-7149-TST-channel1_CTTTCAAGTAGGTACG-1 -0.01895955\n",
       "HTAPP-878-SMP-7149-TST-channel1_CACCGTTGTTCTGACA-1 -0.01895955\n",
       "HTAPP-878-SMP-7149-TST-channel1_GAGTTTGCACAACGTT-1 -0.01895955\n",
       "HTAPP-878-SMP-7149-TST-channel1_GATGACTTCTTTGCTA-1 -0.01895955"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dfScaled[1:5,1:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11901e25",
   "metadata": {},
   "source": [
    "We use the adverSCarial getSignGenes function to get the most significant genes between clusters, ensuring that all pairs of clusters are equivalently representated."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f3803ca2",
   "metadata": {},
   "outputs": [],
   "source": [
    "start_time <- Sys.time()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "9e8a90a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "sign_genes <- getSignGenes(dfScaled, clusters_so)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "8a95bca9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Time difference of 17.79068 mins"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "Sys.time() - start_time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "bd97e2f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>gene</th><th scope=col>pval</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>HSPG2</th><td>HSPG2   </td><td> 0.000000e+00</td></tr>\n",
       "\t<tr><th scope=row>MSR1</th><td>MSR1    </td><td>6.526758e-284</td></tr>\n",
       "\t<tr><th scope=row>PRKG1</th><td>PRKG1   </td><td>3.456779e-180</td></tr>\n",
       "\t<tr><th scope=row>CASZ1</th><td>CASZ1   </td><td> 0.000000e+00</td></tr>\n",
       "\t<tr><th scope=row>ARHGAP15</th><td>ARHGAP15</td><td>2.876970e-183</td></tr>\n",
       "\t<tr><th scope=row>MCAM</th><td>MCAM    </td><td>1.122424e-226</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & gene & pval\\\\\n",
       "  & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\tHSPG2 & HSPG2    &  0.000000e+00\\\\\n",
       "\tMSR1 & MSR1     & 6.526758e-284\\\\\n",
       "\tPRKG1 & PRKG1    & 3.456779e-180\\\\\n",
       "\tCASZ1 & CASZ1    &  0.000000e+00\\\\\n",
       "\tARHGAP15 & ARHGAP15 & 2.876970e-183\\\\\n",
       "\tMCAM & MCAM     & 1.122424e-226\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 2\n",
       "\n",
       "| <!--/--> | gene &lt;chr&gt; | pval &lt;dbl&gt; |\n",
       "|---|---|---|\n",
       "| HSPG2 | HSPG2    |  0.000000e+00 |\n",
       "| MSR1 | MSR1     | 6.526758e-284 |\n",
       "| PRKG1 | PRKG1    | 3.456779e-180 |\n",
       "| CASZ1 | CASZ1    |  0.000000e+00 |\n",
       "| ARHGAP15 | ARHGAP15 | 2.876970e-183 |\n",
       "| MCAM | MCAM     | 1.122424e-226 |\n",
       "\n"
      ],
      "text/plain": [
       "         gene     pval         \n",
       "HSPG2    HSPG2     0.000000e+00\n",
       "MSR1     MSR1     6.526758e-284\n",
       "PRKG1    PRKG1    3.456779e-180\n",
       "CASZ1    CASZ1     0.000000e+00\n",
       "ARHGAP15 ARHGAP15 2.876970e-183\n",
       "MCAM     MCAM     1.122424e-226"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(sign_genes$results)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36a916f6",
   "metadata": {},
   "source": [
    "We launch the attack with the advCGD function with alpha=2 and epsilon=2 parameters, which lead to a big modification on a few genes. You can try with alpha=1 and epsilon=1 to have small modifications of more genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "0adf26d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "start_time <- Sys.time()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "d17b3e12",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "macrophage\n",
      "\n",
      "53126\n",
      "\n",
      "5312\n",
      "\n",
      "392\n",
      "\n",
      "New cluster target: T_cell\n",
      "\n",
      "HSPG2 1\n",
      "\n",
      "Number of original annot macrophage : 370\n",
      "\n",
      " mean 0.959160038401541 delt 0\n",
      "\n",
      "Number of T_cell : 14\n",
      "\n",
      " mean 0.737670271885174 delt 0\n",
      "\n",
      "Number of modified cells 0\n",
      "\n",
      "MSR1 2\n",
      "\n",
      "Number of original annot macrophage : 370\n",
      "\n",
      " mean 0.959160038401541 delt 0\n",
      "\n",
      "Number of T_cell : 14\n",
      "\n",
      " mean 0.737670271885174 delt 0\n",
      "\n",
      "Number of modified cells 378\n",
      "\n",
      "PRKG1 3\n",
      "\n",
      "Number of original annot macrophage : 361\n",
      "\n",
      " mean 0.936429968420805 delt -0.022730069980736\n",
      "\n",
      "Number of T_cell : 16\n",
      "\n",
      " mean 0.762200657720953 delt 0.0245303858357792\n",
      "\n",
      "Number of modified cells 0\n",
      "\n",
      "CASZ1 4\n",
      "\n",
      "Number of original annot macrophage : 361\n",
      "\n",
      " mean 0.936429968420805 delt 0\n",
      "\n",
      "Number of T_cell : 16\n",
      "\n",
      " mean 0.762200657720953 delt 0\n",
      "\n",
      "Number of modified cells 0\n",
      "\n",
      "ARHGAP15 5\n",
      "\n",
      "Number of original annot macrophage : 361\n",
      "\n",
      " mean 0.936429968420805 delt 0\n",
      "\n",
      "Number of T_cell : 16\n",
      "\n",
      " mean 0.762200657720953 delt 0\n",
      "\n",
      "Number of modified cells 376\n",
      "\n",
      "MCAM 6\n",
      "\n",
      "Number of original annot macrophage : 361\n",
      "\n",
      " mean 0.936057724252099 delt -0.000372244168706404\n",
      "\n",
      "Number of T_cell : 16\n",
      "\n",
      " mean 0.767308161684758 delt 0.00510750396380466\n",
      "\n",
      "Number of modified cells 0\n",
      "\n",
      "LRMDA 7\n",
      "\n",
      "Number of original annot macrophage : 361\n",
      "\n",
      " mean 0.936057724252099 delt 0\n",
      "\n",
      "Number of T_cell : 16\n",
      "\n",
      " mean 0.767308161684758 delt 0\n",
      "\n",
      "Number of modified cells 0\n",
      "\n",
      "ZEB2 8\n",
      "\n",
      "Number of original annot macrophage : 361\n",
      "\n",
      " mean 0.936057724252099 delt 0\n",
      "\n",
      "Number of T_cell : 16\n",
      "\n",
      " mean 0.767308161684758 delt 0\n",
      "\n",
      "Number of modified cells 376\n",
      "\n",
      "TIE1 9\n",
      "\n",
      "Number of original annot macrophage : 289\n",
      "\n",
      " mean 0.867339598418616 delt -0.0687181258334827\n",
      "\n",
      "Number of T_cell : 52\n",
      "\n",
      " mean 0.54938150603057 delt -0.217926655654189\n",
      "\n",
      "Number of modified cells 0\n",
      "\n",
      "CALD1 10\n",
      "\n",
      "Number of original annot macrophage : 289\n",
      "\n",
      " mean 0.867339598418616 delt 0\n",
      "\n",
      "Number of T_cell : 52\n",
      "\n",
      " mean 0.54938150603057 delt 0\n",
      "\n",
      "Number of modified cells 0\n",
      "\n",
      "PTPRC 11\n",
      "\n",
      "Number of original annot macrophage : 289\n",
      "\n",
      " mean 0.867339598418616 delt 0\n",
      "\n",
      "Number of T_cell : 52\n",
      "\n",
      " mean 0.54938150603057 delt 0\n",
      "\n",
      "Number of modified cells 340\n",
      "\n",
      "CARMN 12\n",
      "\n",
      "Number of original annot macrophage : 273\n",
      "\n",
      " mean 0.898516343171954 delt 0.0311767447533373\n",
      "\n",
      "Number of T_cell : 74\n",
      "\n",
      " mean 0.496518938307286 delt -0.0528625677232833\n",
      "\n",
      "Number of modified cells 0\n",
      "\n",
      "FYB1 13\n",
      "\n",
      "Number of original annot macrophage : 273\n",
      "\n",
      " mean 0.898516343171954 delt 0\n",
      "\n",
      "Number of T_cell : 74\n",
      "\n",
      " mean 0.496518938307286 delt 0\n",
      "\n",
      "Number of modified cells 0\n",
      "\n",
      "NHSL3 14\n",
      "\n",
      "Number of original annot macrophage : 273\n",
      "\n",
      " mean 0.898516343171954 delt 0\n",
      "\n",
      "Number of T_cell : 74\n",
      "\n",
      " mean 0.496518938307286 delt 0\n",
      "\n",
      "Number of modified cells 0\n",
      "\n",
      "LDB2 15\n",
      "\n",
      "Number of original annot macrophage : 273\n",
      "\n",
      " mean 0.898516343171954 delt 0\n",
      "\n",
      "Number of T_cell : 74\n",
      "\n",
      " mean 0.496518938307286 delt 0\n",
      "\n",
      "Number of modified cells 0\n",
      "\n",
      "CDH6 16\n",
      "\n",
      "Number of original annot macrophage : 273\n",
      "\n",
      " mean 0.898516343171954 delt 0\n",
      "\n",
      "Number of T_cell : 74\n",
      "\n",
      " mean 0.496518938307286 delt 0\n",
      "\n",
      "Number of modified cells 0\n",
      "\n",
      "PLXDC2 17\n",
      "\n",
      "Number of original annot macrophage : 273\n",
      "\n",
      " mean 0.898516343171954 delt 0\n",
      "\n",
      "Number of T_cell : 74\n",
      "\n",
      " mean 0.496518938307286 delt 0\n",
      "\n",
      "Number of modified cells 0\n",
      "\n",
      "TEX41 18\n",
      "\n",
      "Number of original annot macrophage : 273\n",
      "\n",
      " mean 0.898516343171954 delt 0\n",
      "\n",
      "Number of T_cell : 74\n",
      "\n",
      " mean 0.496518938307286 delt 0\n",
      "\n",
      "Number of modified cells 0\n",
      "\n",
      "ST6GALNAC3 19\n",
      "\n",
      "Number of original annot macrophage : 273\n",
      "\n",
      " mean 0.898516343171954 delt 0\n",
      "\n",
      "Number of T_cell : 74\n",
      "\n",
      " mean 0.496518938307286 delt 0\n",
      "\n",
      "Number of modified cells 0\n",
      "\n",
      "CD96 20\n",
      "\n",
      "Number of original annot macrophage : 273\n",
      "\n",
      " mean 0.898516343171954 delt 0\n",
      "\n",
      "Number of T_cell : 74\n",
      "\n",
      " mean 0.496518938307286 delt 0\n",
      "\n",
      "Number of modified cells 318\n",
      "\n",
      "SKAP1 21\n",
      "\n",
      "Number of original annot macrophage : 263\n",
      "\n",
      " mean 0.908513028428304 delt 0.00999668525635067\n",
      "\n",
      "Number of T_cell : 100\n",
      "\n",
      " mean 0.492207172926131 delt -0.00431176538115502\n",
      "\n",
      "Number of modified cells 292\n",
      "\n",
      "CHST11 22\n",
      "\n",
      "Number of original annot macrophage : 259\n",
      "\n",
      " mean 0.911505235828596 delt 0.00299220740029216\n",
      "\n",
      "Number of T_cell : 113\n",
      "\n",
      " mean 0.488722735495377 delt -0.00348443743075405\n",
      "\n",
      "Number of modified cells 279\n",
      "\n",
      "CALCRL 23\n",
      "\n",
      "Number of original annot macrophage : 216\n",
      "\n",
      " mean 0.965654773469977 delt 0.0541495376413802\n",
      "\n",
      "Number of T_cell : 162\n",
      "\n",
      " mean 0.601266669889959 delt 0.112543934394581\n",
      "\n",
      "Number of modified cells 0\n",
      "\n",
      "CYP4Z1 24\n",
      "\n",
      "Number of original annot macrophage : 216\n",
      "\n",
      " mean 0.965654773469977 delt 0\n",
      "\n",
      "Number of T_cell : 162\n",
      "\n",
      " mean 0.601266669889959 delt 0\n",
      "\n",
      "Number of modified cells 0\n",
      "\n",
      "LHFPL6 25\n",
      "\n",
      "Number of original annot macrophage : 216\n",
      "\n",
      " mean 0.965654773469977 delt 0\n",
      "\n",
      "Number of T_cell : 162\n",
      "\n",
      " mean 0.601266669889959 delt 0\n",
      "\n",
      "Number of modified cells 0\n",
      "\n",
      "MACF1 26\n",
      "\n",
      "Number of original annot macrophage : 216\n",
      "\n",
      " mean 0.965654773469977 delt 0\n",
      "\n",
      "Number of T_cell : 162\n",
      "\n",
      " mean 0.601266669889959 delt 0\n",
      "\n",
      "Number of modified cells 0\n",
      "\n",
      "ADGRB3 27\n",
      "\n",
      "Number of original annot macrophage : 216\n",
      "\n",
      " mean 0.965654773469977 delt 0\n",
      "\n",
      "Number of T_cell : 162\n",
      "\n",
      " mean 0.601266669889959 delt 0\n",
      "\n",
      "Number of modified cells 0\n",
      "\n",
      "FMNL2 28\n",
      "\n",
      "Number of original annot macrophage : 216\n",
      "\n",
      " mean 0.965654773469977 delt 0\n",
      "\n",
      "Number of T_cell : 162\n",
      "\n",
      " mean 0.601266669889959 delt 0\n",
      "\n",
      "Number of modified cells 0\n",
      "\n",
      "IGFBP7 29\n",
      "\n",
      "Number of original annot macrophage : 216\n",
      "\n",
      " mean 0.965654773469977 delt 0\n",
      "\n",
      "Number of T_cell : 162\n",
      "\n",
      " mean 0.601266669889959 delt 0\n",
      "\n",
      "Number of modified cells 0\n",
      "\n",
      "ADGRL4 30\n",
      "\n",
      "Number of original annot macrophage : 216\n",
      "\n",
      " mean 0.965654773469977 delt 0\n",
      "\n",
      "Number of T_cell : 162\n",
      "\n",
      " mean 0.601266669889959 delt 0\n",
      "\n",
      "Number of modified cells 0\n",
      "\n",
      "RIPOR2 31\n",
      "\n",
      "Number of original annot macrophage : 216\n",
      "\n",
      " mean 0.965654773469977 delt 0\n",
      "\n",
      "Number of T_cell : 162\n",
      "\n",
      " mean 0.601266669889959 delt 0\n",
      "\n",
      "Number of modified cells 230\n",
      "\n",
      "SLIT3 32\n",
      "\n",
      "Number of original annot macrophage : 197\n",
      "\n",
      " mean 0.977875943865762 delt 0.0122211703957856\n",
      "\n",
      "Number of T_cell : 182\n",
      "\n",
      " mean 0.630105582082103 delt 0.0288389121921443\n",
      "\n",
      "Number of modified cells 0\n",
      "\n",
      "PTPRG 33\n",
      "\n",
      "Number of original annot macrophage : 197\n",
      "\n",
      " mean 0.977875943865762 delt 0\n",
      "\n",
      "Number of T_cell : 182\n",
      "\n",
      " mean 0.630105582082103 delt 0\n",
      "\n",
      "Number of modified cells 0\n",
      "\n",
      "SAMSN1 34\n",
      "\n",
      "Number of original annot macrophage : 197\n",
      "\n",
      " mean 0.977875943865762 delt 0\n",
      "\n",
      "Number of T_cell : 182\n",
      "\n",
      " mean 0.630105582082103 delt 0\n",
      "\n",
      "Number of modified cells 0\n",
      "\n",
      "PTPRF 35\n",
      "\n",
      "Number of original annot macrophage : 197\n",
      "\n",
      " mean 0.977875943865762 delt 0\n",
      "\n",
      "Number of T_cell : 182\n",
      "\n",
      " mean 0.630105582082103 delt 0\n",
      "\n",
      "Number of modified cells 0\n",
      "\n",
      "DOCK10 36\n",
      "\n",
      "Number of original annot macrophage : 197\n",
      "\n",
      " mean 0.977875943865762 delt 0\n",
      "\n",
      "Number of T_cell : 182\n",
      "\n",
      " mean 0.630105582082103 delt 0\n",
      "\n",
      "Number of modified cells 0\n",
      "\n",
      "VCAN 37\n",
      "\n",
      "Number of original annot macrophage : 197\n",
      "\n",
      " mean 0.977875943865762 delt 0\n",
      "\n",
      "Number of T_cell : 182\n",
      "\n",
      " mean 0.630105582082103 delt 0\n",
      "\n",
      "Number of modified cells 0\n",
      "\n",
      "NEAT1 38\n",
      "\n",
      "Number of original annot macrophage : 197\n",
      "\n",
      " mean 0.977875943865762 delt 0\n",
      "\n",
      "Number of T_cell : 182\n",
      "\n",
      " mean 0.630105582082103 delt 0\n",
      "\n",
      "Number of modified cells 0\n",
      "\n",
      "ITGA1 39\n",
      "\n",
      "Number of original annot macrophage : 197\n",
      "\n",
      " mean 0.977875943865762 delt 0\n",
      "\n",
      "Number of T_cell : 182\n",
      "\n",
      " mean 0.630105582082103 delt 0\n",
      "\n",
      "Number of modified cells 0\n",
      "\n",
      "ARHGAP29 40\n",
      "\n",
      "Number of original annot macrophage : 197\n",
      "\n",
      " mean 0.977875943865762 delt 0\n",
      "\n",
      "Number of T_cell : 182\n",
      "\n",
      " mean 0.630105582082103 delt 0\n",
      "\n",
      "Number of modified cells 0\n",
      "\n",
      "SOX5 41\n",
      "\n",
      "Number of original annot macrophage : 197\n",
      "\n",
      " mean 0.977875943865762 delt 0\n",
      "\n",
      "Number of T_cell : 182\n",
      "\n",
      " mean 0.630105582082103 delt 0\n",
      "\n",
      "Number of modified cells 0\n",
      "\n",
      "FYN 42\n",
      "\n",
      "Number of original annot macrophage : 197\n",
      "\n",
      " mean 0.977875943865762 delt 0\n",
      "\n",
      "Number of T_cell : 182\n",
      "\n",
      " mean 0.630105582082103 delt 0\n",
      "\n",
      "Number of modified cells 210\n",
      "\n"
     ]
    }
   ],
   "source": [
    "igd4c_results <- advCGD(so_pbmc_test, clusters_so, 'macrophage',\n",
    "                          scAnnotatR_classifier, alpha=2,\n",
    "                          epsilon=2, slot=\"scale.data\",\n",
    "                          genes=sign_genes$results$gene[1:500],\n",
    "                         verbose=T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ace92c46",
   "metadata": {},
   "source": [
    "# Computation time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "12d68a4e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Time difference of 21.19147 mins"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "Sys.time() - start_time"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5896fe8",
   "metadata": {},
   "source": [
    "We can see the genes that have been modified"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "7b919c8f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'MSR1'</li><li>'ARHGAP15'</li><li>'ZEB2'</li><li>'PTPRC'</li><li>'CD96'</li><li>'SKAP1'</li><li>'CHST11'</li><li>'RIPOR2'</li><li>'FYN'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'MSR1'\n",
       "\\item 'ARHGAP15'\n",
       "\\item 'ZEB2'\n",
       "\\item 'PTPRC'\n",
       "\\item 'CD96'\n",
       "\\item 'SKAP1'\n",
       "\\item 'CHST11'\n",
       "\\item 'RIPOR2'\n",
       "\\item 'FYN'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'MSR1'\n",
       "2. 'ARHGAP15'\n",
       "3. 'ZEB2'\n",
       "4. 'PTPRC'\n",
       "5. 'CD96'\n",
       "6. 'SKAP1'\n",
       "7. 'CHST11'\n",
       "8. 'RIPOR2'\n",
       "9. 'FYN'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"MSR1\"     \"ARHGAP15\" \"ZEB2\"     \"PTPRC\"    \"CD96\"     \"SKAP1\"    \"CHST11\"  \n",
       "[8] \"RIPOR2\"   \"FYN\"     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "igd4c_results$modGenes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8dd443b1",
   "metadata": {},
   "source": [
    "And the modified Seurat object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "3461ddcb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "An object of class Seurat \n",
       "25345 features across 5312 samples within 1 assay \n",
       "Active assay: RNA (25345 features, 0 variable features)\n",
       " 3 layers present: counts, data, scale.data"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "igd4c_results$expr"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6893b0e6",
   "metadata": {},
   "source": [
    "The malignant cell cluster is now classified as CD8 T cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "80e16f78",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'T_cell'"
      ],
      "text/latex": [
       "'T\\_cell'"
      ],
      "text/markdown": [
       "'T_cell'"
      ],
      "text/plain": [
       "[1] \"T_cell\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "new_classification <- scAnnotatR_classifier(igd4c_results$expr, clusters_so, 'macrophage')\n",
    "new_classification$prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c5a24b2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a52ca7b9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "e0762f1b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.0 (2023-04-21)\n",
       "Platform: x86_64-pc-linux-gnu (64-bit)\n",
       "Running under: Ubuntu 22.04.1 LTS\n",
       "\n",
       "Matrix products: default\n",
       "BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0 \n",
       "LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=fr_FR.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=fr_FR.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=fr_FR.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=fr_FR.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Europe/Paris\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       "[1] stats4    stats     graphics  grDevices utils     datasets  methods  \n",
       "[8] base     \n",
       "\n",
       "other attached packages:\n",
       " [1] adverSCarial_1.3.6          IRdisplay_1.1              \n",
       " [3] scAnnotatR_1.8.0            SingleCellExperiment_1.23.0\n",
       " [5] SummarizedExperiment_1.31.1 Biobase_2.61.0             \n",
       " [7] GenomicRanges_1.53.1        GenomeInfoDb_1.37.2        \n",
       " [9] IRanges_2.35.2              S4Vectors_0.39.1           \n",
       "[11] BiocGenerics_0.47.0         MatrixGenerics_1.13.0      \n",
       "[13] matrixStats_1.0.0           Seurat_5.1.0               \n",
       "[15] SeuratObject_5.0.2          sp_2.0-0                   \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] RcppAnnoy_0.0.20              splines_4.3.0                \n",
       "  [3] later_1.3.1                   pbdZMQ_0.3-9                 \n",
       "  [5] bitops_1.0-7                  filelock_1.0.2               \n",
       "  [7] tibble_3.2.1                  polyclip_1.10-4              \n",
       "  [9] hardhat_1.3.1                 pROC_1.18.5                  \n",
       " [11] rpart_4.1.19                  fastDummies_1.7.4            \n",
       " [13] lifecycle_1.0.3               globals_0.16.2               \n",
       " [15] lattice_0.21-8                MASS_7.3-60                  \n",
       " [17] magrittr_2.0.3                plotly_4.10.2                \n",
       " [19] yaml_2.3.7                    httpuv_1.6.11                \n",
       " [21] sctransform_0.4.1             spam_2.11-0                  \n",
       " [23] spatstat.sparse_3.0-2         reticulate_1.36.1            \n",
       " [25] cowplot_1.1.1                 pbapply_1.7-2                \n",
       " [27] DBI_1.1.3                     RColorBrewer_1.1-3           \n",
       " [29] lubridate_1.9.3               abind_1.4-5                  \n",
       " [31] zlibbioc_1.47.0               Rtsne_0.16                   \n",
       " [33] purrr_1.0.1                   RCurl_1.98-1.12              \n",
       " [35] nnet_7.3-19                   rappdirs_0.3.3               \n",
       " [37] ipred_0.9-14                  lava_1.8.0                   \n",
       " [39] GenomeInfoDbData_1.2.10       data.tree_1.1.0              \n",
       " [41] ggrepel_0.9.3                 irlba_2.3.5.1                \n",
       " [43] listenv_0.9.0                 spatstat.utils_3.0-3         \n",
       " [45] goftest_1.2-3                 RSpectra_0.16-2              \n",
       " [47] spatstat.random_3.1-5         fitdistrplus_1.1-11          \n",
       " [49] parallelly_1.36.0             leiden_0.4.3                 \n",
       " [51] codetools_0.2-19              DelayedArray_0.27.5          \n",
       " [53] tidyselect_1.2.0              BiocFileCache_2.9.0          \n",
       " [55] base64enc_0.1-3               spatstat.explore_3.2-1       \n",
       " [57] jsonlite_1.8.5                caret_6.0-94                 \n",
       " [59] e1071_1.7-14                  ellipsis_0.3.2               \n",
       " [61] progressr_0.13.0              iterators_1.0.14             \n",
       " [63] ggridges_0.5.4                survival_3.5-5               \n",
       " [65] foreach_1.5.2                 tools_4.3.0                  \n",
       " [67] ica_1.0-3                     Rcpp_1.0.10                  \n",
       " [69] glue_1.6.2                    prodlim_2023.08.28           \n",
       " [71] gridExtra_2.3                 SparseArray_1.1.10           \n",
       " [73] dplyr_1.1.2                   withr_2.5.0                  \n",
       " [75] BiocManager_1.30.21           fastmap_1.1.1                \n",
       " [77] fansi_1.0.4                   digest_0.6.32                \n",
       " [79] timechange_0.3.0              R6_2.5.1                     \n",
       " [81] mime_0.12                     colorspace_2.1-0             \n",
       " [83] scattermore_1.2               tensor_1.5                   \n",
       " [85] spatstat.data_3.0-1           RSQLite_2.3.1                \n",
       " [87] utf8_1.2.3                    tidyr_1.3.0                  \n",
       " [89] generics_0.1.3                recipes_1.0.10               \n",
       " [91] data.table_1.14.8             class_7.3-22                 \n",
       " [93] httr_1.4.6                    htmlwidgets_1.6.2            \n",
       " [95] S4Arrays_1.1.4                ModelMetrics_1.2.2.2         \n",
       " [97] uwot_0.1.15                   pkgconfig_2.0.3              \n",
       " [99] gtable_0.3.3                  timeDate_4032.109            \n",
       "[101] blob_1.2.4                    lmtest_0.9-40                \n",
       "[103] XVector_0.41.1                htmltools_0.5.5              \n",
       "[105] dotCall64_1.2                 scales_1.2.1                 \n",
       "[107] png_0.1-8                     gower_1.0.1                  \n",
       "[109] reshape2_1.4.4                uuid_1.1-0                   \n",
       "[111] nlme_3.1-162                  curl_5.0.1                   \n",
       "[113] proxy_0.4-27                  repr_1.1.6                   \n",
       "[115] zoo_1.8-12                    cachem_1.0.8                 \n",
       "[117] stringr_1.5.0                 BiocVersion_3.18.0           \n",
       "[119] KernSmooth_2.23-21            parallel_4.3.0               \n",
       "[121] miniUI_0.1.1.1                AnnotationDbi_1.63.1         \n",
       "[123] pillar_1.9.0                  grid_4.3.0                   \n",
       "[125] vctrs_0.6.3                   RANN_2.6.1                   \n",
       "[127] promises_1.2.0.1              dbplyr_2.3.2                 \n",
       "[129] xtable_1.8-4                  cluster_2.1.4                \n",
       "[131] evaluate_0.21                 cli_3.6.1                    \n",
       "[133] compiler_4.3.0                rlang_1.1.1                  \n",
       "[135] crayon_1.5.2                  future.apply_1.11.0          \n",
       "[137] plyr_1.8.8                    stringi_1.7.12               \n",
       "[139] viridisLite_0.4.2             deldir_1.0-9                 \n",
       "[141] Biostrings_2.69.1             munsell_0.5.0                \n",
       "[143] lazyeval_0.2.2                spatstat.geom_3.2-1          \n",
       "[145] Matrix_1.6-4                  IRkernel_1.3.2               \n",
       "[147] RcppHNSW_0.4.1                patchwork_1.1.2              \n",
       "[149] bit64_4.0.5                   future_1.32.0                \n",
       "[151] ggplot2_3.4.2                 KEGGREST_1.41.0              \n",
       "[153] shiny_1.7.4                   interactiveDisplayBase_1.39.0\n",
       "[155] AnnotationHub_3.9.1           kernlab_0.9-32               \n",
       "[157] ROCR_1.0-11                   igraph_1.5.0                 \n",
       "[159] memoise_2.0.1                 bit_4.0.5                    \n",
       "[161] ape_5.8                      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5457df0e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "273f6f43",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
